hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomRoyaltyFeeAssessor.java

```diff
@@ -0,0 +1,147 @@
+        final var exchangedValue = getFungibleCredits(result, tokenId, sender);
+        for (final var fee : feeMeta.customFees()) {
+            final var collector = fee.feeCollectorAccountId();
+            if (!fee.fee().kind().equals(CustomFee.FeeOneOfType.ROYALTY_FEE)) {
+                continue;
+            }
+            final var royaltyFee = fee.royaltyFeeOrThrow();
+            // If there are no fungible units to the receiver, then  if there is a fallback fee
+            // then receiver should pay the fallback fee
+            if (exchangedValue.isEmpty()) {
+                if (!royaltyFee.hasFallbackFee()) {
+                    continue;
+                }
+                final var fallback = royaltyFee.fallbackFeeOrThrow();
+                // A NFT transfer with royalty fees to an unknown alias is not possible, since
+                // the auto-created account will not have any hbar to pay the fallback fee
+                // Validate the account balance is greater than zero
+                validateTrue(
+                        accountStore.get(receiver).tinybarBalance() != 0,
+                        INSUFFICIENT_SENDER_ACCOUNT_BALANCE_FOR_CUSTOM_FEE);
+                final var fallbackFee = asFixedFee(
+                        fallback.amount(), fallback.denominatingTokenId(), collector, fee.allCollectorsAreExempt());
+                fixedFeeAssessor.assessFixedFee(feeMeta, receiver, fallbackFee, result);
+            } else {
+                if (!isPayerExempt(feeMeta, fee, sender)) {
+                    chargeRoyalty(exchangedValue, feeMeta, fee, result);
+                }
+            }
+        }
+    }
+
+    private void chargeRoyalty(
+            @NonNull Map<AccountID, Pair<Long, TokenID>> exchangedValues,
+            @NonNull final CustomFeeMeta feeMeta,
+            @NonNull final CustomFee fee,
+            @NonNull final AssessmentResult result) {
+        for (final var exchange : exchangedValues.entrySet()) {
+            final var account = exchange.getKey();
+            final var value = exchange.getValue();
+            final var amount = value.getLeft();
+            final var tokenId = value.getRight();
+
+            final var royaltySpec = fee.royaltyFeeOrThrow();
+            final var feeCollector = fee.feeCollectorAccountIdOrThrow();
+
+            final var royalty = safeFractionMultiply(
+                    royaltySpec.exchangeValueFraction().numerator(),
+                    royaltySpec.exchangeValueFraction().denominator(),
+                    amount);
+            validateTrue(royalty <= amount, INSUFFICIENT_SENDER_ACCOUNT_BALANCE_FOR_CUSTOM_FEE);
+
+            final var denom = tokenId == null ? null : tokenId;
+            /* The id of the charging token is only used here to avoid recursively charging
+            on fees charged in the units of their denominating token; but this is a credit,
+            hence the id is irrelevant and we can use MISSING_ID. */
```

Comment:
Where is `MISSING_ID` used in this code?

Reply:
Removed